Automattic/jetpack
# Disable Sync sending on Backup API Requests

Jetpack Backup/Rewind makes API calls to Jetpack Sites to get current content. These requests are triggered when Jetpack Sync sends data to WordPress.com. Recently there have been concerns with unexpected code flows in plugins creating new posts on every web request. When this happens we end up in infinite loops of the following nature.

Visitor to Jetpack powered Site -> create Post -> Sync to WP.com -> WP.com makes API call to Jetpack powered Site -> create Post -> Sync to WP.com -> .... 

Fixes https://wp.me/p7H4VZ-2FT

#### Changes proposed in this Pull Request:
This patch 

1) introduces a new constant JETPACK_SYNC_READ_ONLY that will stop sync as the request has explicitly stated its intention is read only and shouldn't be generating content.

2) Uses this constant in backup APIs so they don't trigger Sync

3) Unit Tests to ensure sync does not send and of the helper function mark_sync_read_only in Actions class

Now if a plugin incorrectly generates a post on every request we will not infinitely make requests but instead the flow will be

Visitor to Jetpack Site -> create Post -> Sync to WP.com -> WP.com makes API call to Jetpack Site -> create Post -> /end

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:

* Checkout Master Branch
* Ensure Jetpack is connected and sync is enabled
* Add below snippet so that every request creates a new Post :scream:
```
function local_init_create_post() {
	$blogusers = get_users( [ 'role__in' => [ 'administrator'] ] );
	$my_post = array(
		'post_title'    => "Spam Post - cleanup",
		'post_content'  => "Beautiful Spam",
		'post_status'   => 'publish',
		'post_author'   => $blogusers[0]->ID
	);

	wp_insert_post( $my_post );
}
add_filter( 'init', 'local_init_create_post' );
```
* Visit your website (front end request)
* Verify that new "Spam Post - cleanup" posts are created (shell, mysql or kibana view of sync actions) and continue to be created without any additional views of your site.
* comment out the add_filter( 'init') line and wait for a few minutes for posts and api requests to calm down. 
* checkout sync/ixr-disable-rewind branch
* un comment out the add_filter line
* Visit your webiste (FE request)
* Verify that creation of new "Spam Post.- cleanup" post occur
* Verify that creation stops when no traffic/activity

We are confirming that backup/sync stops its infinite request loop. We are not stopping the misbehaving code from creating posts on every request

Don't forget to remove the test code and clean your posts table when done with testing :whistling: 

#### Proposed changelog entry for your changes:
* Jetpack Sync :: Introduce new JETPACK_SYNC_READ_ONLY constant that can be used to disable sync from sending data in the current request. Apply JETPACK_SYNC_READ_ONLY to backup requests.
